django/db/models/sql/query.py

```diff
@@ -2078,3 +2078,15 @@ def update_join_types(self, query):
         query.promote_joins(to_promote)
         query.demote_joins(to_demote)
         return to_demote
+
+
+class ExpressionIndexQuery:
+
+    def __init__(self, model):
+        self.model = model
+
+    def resolve_ref(self, name, *args, **kwargs):
+        if not any(f.name == name for f in self.model._meta.concrete_fields):
+            raise ValueError("Invalid reference to field '%s' on model '%s'" % (name, self.model._meta.label))
+        source_field = self.model._meta.get_field(name)
+        return Col(None, source_field, index_col=True)  # None refers to the table alias we don't have here
```

Comment:
Any thoughts about using a `Col` subclass here instead of this flag? You'd simply have to `return IndexCol(None, source_field)`.

Reply:
I had a very similar requirement in my work on Check Constraints: https://github.com/django/django/commit/1a5ade0c3b55fb6038c5941787212cb2fe079d6d#diff-fe637a3ec6af9cbe2ee87293000e7c82R751

It's probably worthwhile checking if we can share some implementation details here.